aip/client-libraries/4234.md

```diff
@@ -0,0 +1,103 @@
+targets are named, clients can have a uniform experience from API to API.
+
+Some of the various language constraints are as follows:
+
+*   Java's `option java_outer_classname`: per the API style guide is the name of
+    the proto file, with "Proto" appended. Unless `option java_multiple_files` is true,
+    the file name defines the packaging of the types.
+*   Go's target names will define the import path for the result. However, Go's package
+    names have a more restrictive character set (e.g., no dashes), and in some cases, that
+    means that the import path and package name will disagree. This interferes with some
+    tooling.
+
+## Guidance
+
+### Targets
+
+In the following file name cases, if the service name or resource name is more than one
+word, they should be separated by underscores.
+
+#### Services and request/response messages
+
+Individual services along with their request and response messages **should** live in
+separate `proto_library` targets, named `{service}_service_proto`.
+
+**NOTE:** Services **should** not share request or response messages. Reuse of resources
+or common messages is preferred.
+    
+#### Resource messages
+
+*   Resources that are a primary resource for an individual service, **should** be 
+    in a `proto_library` target named `{resource}_proto`.
+*   Otherwise, per-service resources **should** be in a `proto_library` target named 
+    `{service}_resources_proto`.
+*   If there are remaining shared resources, they **should** be in a `proto_library`
+    target named `resources_proto`.
+    
+#### Additional messages
+
+Remaining necessary messages **should** be in a `proto_library` target named
+`common_proto`.
+
+### `.proto` file naming convention
+
+In the case where there's a single `.proto` file in the `srcs` of the
+`proto_library`, it should be named by replacing the `_proto` suffix with `.proto`.
+
+### `{lang}_proto_library` naming convention
+
+The per-language rules, e.g., `cc_proto_library`, **should** be named by replacing the
+`_proto` suffix on the `proto_library` target with `_{lang}_proto`. e.g.,
+
+```
+proto_library(name = "foo_proto", …)
+cc_proto_library(name = "foo_cc_proto", deps = [":foo_proto"], …)
+```
+
+### `{lang}_{rpc}_library` naming convention
+
+The per-language, per-RPC rules, e.g., `cc_grpc_library`, **should** be named by replacing
+the `_proto` suffix on the `{lang}_proto_library` target with `_{rpc}`. e.g.,
+
+```
+cc_proto_library(name = "bar_service_cc_proto", deps = [":bar_service_proto"], …)
+cc_grpc_library(
+    name = "bar_service_cc_grpc",
+    srcs = [":bar_service_proto"],
+    deps = [":bar_service_cc_proto"],
+    …
+)
+```
+
+### Adding services
+
+If in the course of an alpha or beta period for an API, one or more new services are added
+or resources are made shared or specific such that they'd belong to a different or new
+`proto_library` target,
+
+*   Prefer incrementing the _Y_ of vXalphaY to a new version and making the changes
```

Comment:
This will become more or less impossible with where our versioning scheme is likely to go.

Reply:
I would take pointers to that. If this is going to be a dead-end, it could be removed.

Reply:
The basic idea is going to be that we have a single proto for a major version, and then we have an enum selector for ALPHA, BETA, STABLE, and then have post-processing to provide "views" on the protos.

We have not written a document on GitHub yet because there are still a bunch of internal-baseball questions to work out.

Reply:
This part definitely needs to be removed. :-)